src/Embedders/Word2Vec.php

```diff
@@ -0,0 +1,924 @@
+            }
+        }
+
+        return array_slice($result, 0, $top, true);
+    }
+
+    /**
+     * Returns the word embedding for a given word.
+     *
+     * @param string $word
+     * @param bool $useNorm
+     * @return \Tensor\Vector|null $result
+     */
+    public function wordVec(string $word, bool $useNorm = true) : ?Vector
+    {
+        if (!array_key_exists($word, $this->vocab)) {
+            return null;
+        }
+
+        if ($useNorm) {
+            $result = $this->vectorsNorm[$this->vocab[$word]['index']];
+        } else {
+            $result = $this->vectors[$this->vocab[$word]['index']];
+        }
+
+        return $result;
+    }
+
+    /**
+     * Returns the word embedding, or a vector of zeros if empty, for a given word.
+     *
+     * @param string $word
+     * @param bool $useNorm
+     * @return \Tensor\Vector $result
+     */
+    public function embedWord(string $word, bool $useNorm = true) : Vector
+    {
+        $wordEmbedding = $this->wordVec($word);
+
+        if (!$wordEmbedding) {
+            $wordEmbedding = Vector::zeros($this->dimensions);
+        }
+
+        return $wordEmbedding;
+    }
+
+    /**
+     * Scanning vocab, preparing vocab, sorting vocab, setting sampling methods
+     *
+     * @param string[] $sentences
+     */
+    private function preprocess(array $sentences) : void
+    {
+        $words = $this->prepCorpus($sentences);
+
+        $this->scanVocab($words);
+        $this->prepVocab();
+        $this->sortVocab();
+
+        switch ($this->layer) {
+            case 'hs':
+                $this->createBinaryTree();
+                $this->trainMethod = 'trainPairSgHS';
+                break;
+            case 'neg':
+                $this->createCumTable();
+                $this->trainMethod = 'trainPairSgNeg';
+                break;
+        }
+    }
+
+    /**
+     * Parsing, sanitizing, & preparing array of sentences to generate & set corpus.
+     *
+     * @param string[] $sentences
+     * @return string[] $words
+     */
+    private function prepCorpus(array $sentences) : array
+    {
+        $words = [];
+
+        foreach ($sentences as $sentence) {
+            $preppedSentence = $this->prepSentence($sentence);
+
+            $this->corpus[] = $preppedSentence;
+            $words = array_merge($words, $preppedSentence);
+        }
+
+        return $words;
+    }
+
+    /**
+     * Sanitizing & exploding a provided sentence.
+     *
+     * @param string $sentence
+     * @return string[] $exploded
+     */
+    private function prepSentence(string $sentence) : array
+    {
+        $sentence = (string) preg_replace('#[[:punct:]]#', '', strtolower($sentence));
+        $sentence = (string) preg_replace('/\s\s+/', ' ', str_replace("\n", ' ', $sentence));
+        $sentence = trim($sentence);
+        return explode(' ', $sentence);
+    }
+
+    /**
+     * Preparing and setting an initial raw vocab array for additional preprocessing
```

Comment:
Not a big deal but we tend to stay away from participles in our comments and just stick to the verb. Example 'Prepare and set an initial raw vocab ...'